### PR TITLE
[Parse] Discriminate local variables

### PR DESCRIPTION
**Pull request by rintaro**
Set local discriminator for all local `VarDecl`s. Otherwise, they cannot be discriminated with USRs. This change is needed for rename refactoring which uses USR for discrimiating variable names.

[SR-7205](https://bugs.swift.org/browse/SR-7205), rdar://problem/34701880

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: test/refactoring/rename/Outputs/local/catch_1.swift.expected [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    break
+  }
+}
+
+struct Err1 : Error { }
+func test4(arg: () throws -> Void) {
+  do {
+    try arg()
+  } catch let x as Err1 {
+    print(xRenamed)
```
Comment (rintaro) [inline test/refactoring/rename/Outputs/local/catch_1.swift.expected]: `let x` above should be renamed as well. Investigating.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/IDE/SwiftSourceDocInfo.cpp [hunk 2]
```diff
@@ -581,7 +581,7 @@ std::pair<bool, Pattern*> NameMatcher::walkToPatternPre(Pattern *P) {
   if (isDone() || shouldSkip(P->getSourceRange()))
     return std::make_pair(false, P);
 
-  tryResolve(ASTWalker::ParentTy(P), P->getLoc());
+  tryResolve(ASTWalker::ParentTy(P), P->getStartLoc());
```
Comment (rintaro) [inline lib/IDE/SwiftSourceDocInfo.cpp]: @nkcsgexi is this OK?

**Comment by rintaro** [comment]
@rjmccall In commit message of 892a90033283dd9b3a48173803d6683b7a6143dd, you explicitly excluded physical vars

>    Track a discriminator for named declarations (except physical vars) in local contexts, for future use in mangling.

What was the reason?

**Comment by rjmccall** [comment]
Computing and storing a discriminator means time and memory overhead, and we don't actually need a discriminator because an ordinary local variable doesn't require the creation of anything outside the code-generation of the function.

**Comment by rjmccall** [comment]
If you're really going to attach discriminators to every single local variable, the side-table storage approach is probably unreasonable.

File: lib/IDE/SwiftSourceDocInfo.cpp (hunk 2 above)
Comment (nkcsgexi) [reply lib/IDE/SwiftSourceDocInfo.cpp]: I think this is fine for patterns. We need to be careful for decls though, sometimes `getLoc()` and `getStartLoc` point to different position IIRC.

**Comment by jrose-apple** [comment]
I see you removing the local discriminator from serialization completely. Is it really not needed?

**Comment by rintaro** [comment]
@jrose-apple I believe it's not needed because local discriminators are serialized/deserialized as a part of Decl emission for every `ValueDecl`s. https://github.com/apple/swift/blob/db423c32821a7223a13cdf4786204cf843aa2459/lib/Serialization/Serialization.cpp#L2662-L2667

**Comment by jrose-apple** [comment]
Ah, I didn't realize that was already there! Okay then. Please bump the version number in ModuleFormat.h too!

**Comment by rjmccall** [comment]
Do we really only have 3700 local variable declarations in the stdlib?  That seems extremely low.

Are we not setting a discriminator on parameters?  I guess maybe we don't need to because we can identify them reliably relative to the function they're parameters of; that would definitely keep the count down.

**Comment by rintaro** [comment]
Ah, I forgot about `ParamDecl`. And yes, we need to discriminate them.
```swift
func foo(_ arg: Int) {
  var arg = arg;
}
```
Updated the PR.
Now that, 10492 out of 12368 `VarDecl`s are in local context.

As a side note, this PR impacts the *parse* time by around 1% (0.654s -> 0.662s) and max rss by around 1% (165,376,000 -> 167,411,712).

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rjmccall** [comment]
Are you sure you can't distinguish ParamDecls implicitly?  A ParamDecl should always be uniquely identified by (1) the function it's a parameter of, (2) the fact that it's a parameter, and (3) its index.

**Comment by rintaro** [comment]
Distinguishing ParamDecls without using LocalDiscriminator is technically feasible, but I think using LocalDiscriminator is easiest/convenient way to achieve that. And, as you said, we already have facility for that.
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@rjmccall Could you approve this?

**Review by rjmccall — APPROVED** [review]
This PR looks good to me, but I think you should consider adding some way to test local discriminators more directly.  Maybe a testing switch that prints declaration names with their line, column, and discriminator?

**Comment by rintaro** [comment]
Thanks. That is actually a good idea!

**Comment by rintaro** [comment]
Resolved conflicts.
@swift-ci Please smoke test

**Pull request merged**